File: $_algotsu-kaizen/problemo-set/number-of-steps-to-reduce-a-number-in-binary-representation-to-one.js
/**
 * 1404. Number of Steps to Reduce a Number in Binary Representation to One
 * https://leetcode.com/problems/number-of-steps-to-reduce-a-number-in-binary-representation-to-one/description/
 */

/**
 * @param {string} s
 * @return {number}
 */
var numSteps = function(s) {
    let steps = 0

    /* // Dumb Shit
    let n = parseInt(s, 2)
    while (n != 1) {
        let _ = (n % 2) ? n += 1 : n /= 2 
        ++steps
    }
    */

    let str = new String(s)
    let arr = str.split('')
    let pointer = str.length - 1
    while (pointer > 0) {
        let char = arr[pointer]
        if (char == '1') {
            ++steps
            while (pointer > 0 && arr[pointer] == '1') { 
                ++steps
                --pointer
             }

             if (pointer == 0) ++steps

             arr[pointer] = '1'
        } else {
            ++steps
            --pointer            
        }
    }

    return steps
};

export { numSteps }

console.log("$#", numSteps("1101"));

File: $_algotsu-kaizen/problemo-set/number-of-steps-to-reduce-a-number-in-binary-representation-to-one.c
int numSteps(char* s) {
    // uint8_t buff[0x40] = { 0x0 };
    // uint8_t byte = 0x0;

    char _;
    unsigned short len = 0x0;
    while ((_ = s[len]) && ++len);

    int steps = 0x0;
    int pointer = len - 0x1;

    while (pointer > 0x0)
        if (s[pointer] == '1') {
            ++steps;
            while (pointer > 0x0 && s[pointer] == '1') {
                ++steps;
                --pointer;
             }
             if (pointer == 0x0) ++steps;
             s[pointer] = '1';
        } else {
            ++steps;
            --pointer;
        }

    return steps;
}
